process_pr.py

```diff
@@ -1459,35 +1464,123 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
         ):
             if not dryRun:
                 issue.create_comment(
-                    f"This PR contains too many commits ({pr.commits} > {MAX_INITIAL_COMMITS_IN_PR}). "
+                    "This PR contains too many commits ({0} > {1}). "
                     "Make sure you chose the right target branch.\n"
-                    f"{', '.join([gh_user_char + name for name in CMSSW_ISSUES_TRACKERS])}, "
-                    "you can override this check with `+commit-count`."
+                    "{2}, "
+                    "you can override this check with `+commit-count`.".format(
+                        pr.commits,
+                        MAX_INITIAL_COMMITS_IN_PR,
+                        ", ".join([gh_user_char + name for name in CMSSW_ISSUES_TRACKERS]),
+                    )
                 )
             return
 
         print("Processing commits")
+        commit_cache = {k: v for k, v in bot_cache.items() if REGEX_COMMIT_SHA.match(k)}
+        if commit_cache:
+            new_pr = False
         if pr.commits < MAX_INITIAL_COMMITS_IN_PR or ok_too_many_commits:
+            seen_commits_cnt = sum((not x.get("squashed", False)) for x in commit_cache.values())
+            if all_commits.totalCount < seen_commits_cnt:
+                print(
+                    "Number of commits in cache ({0}) is greater than the number of commits in PR ({1}) - possible squash detected".format(
+                        seen_commits_cnt, all_commits.totalCount
+                    )
+                )
+
+                last_seen_commit_sha = sorted(
+                    [
+                        (k, commit_cache[k]["time"])
+                        for k in commit_cache
+                        if not commit_cache[k].get("squashed", False)
+                    ],
+                    key=lambda q: q[1],
+                )[-1][0]
+                new_head_commit_sha = pr.head.sha
+                base_commit_sha = pr.base.sha
+
+                changes_before_squash = repo.compare(base_commit_sha, last_seen_commit_sha)
+                changes_after_squash = repo.compare(base_commit_sha, new_head_commit_sha)
+
+                diff_before_squash = {
+                    (file.filename, file.patch) for file in changes_before_squash.files
+                }
+                diff_after_squash = {
+                    (file.filename, file.patch) for file in changes_after_squash.files
+                }
+
+                if diff_before_squash ^ diff_after_squash:
+                    print("PR diff changed, will not preserve signatures")
+                else:
+                    print("PR diff not changed, preserving signatures and commit statuses")
+                    bot_cache[new_head_commit_sha] = {
+                        "time": bot_cache[last_seen_commit_sha]["time"] + 1,
```

Comment:
@iarspider , any specific reasons to not use the `time` as it is (instead of adding `1`)?